src/bare_item.rs

```diff
@@ -0,0 +1,410 @@
+    /// ```
+    /// # use sfv::BareItem;
+    /// # use std::convert::TryInto;
+    /// # fn main() -> Result<(), &'static str> {
+    /// let bare_item = BareItem::String("foo".to_owned().try_into()?);
+    /// assert_eq!(bare_item.as_str().unwrap(), "foo");
+    /// Ok(())
+    /// # }
+    /// ```
+    pub fn as_str(&self) -> Option<&str> {
+        match *self {
+            BareItem::String(ref val) => Some(val),
+            _ => None,
+        }
+    }
+    /// If `BareItem` is a `ByteSeq`, returns `&Vec<u8>`, otherwise returns `None`.
+    /// ```
+    /// # use sfv::BareItem;
+    /// let bare_item = BareItem::ByteSeq("foo".to_owned().into_bytes().into());
+    /// assert_eq!(bare_item.as_byte_seq().unwrap().as_slice(), "foo".as_bytes());
+    /// ```
+    pub fn as_byte_seq(&self) -> Option<&Vec<u8>> {
+        match *self {
+            BareItem::ByteSeq(ref val) => Some(&val.0),
+            _ => None,
+        }
+    }
+    /// If `BareItem` is a `Boolean`, returns `bool`, otherwise returns `None`.
+    /// ```
+    /// # use sfv::{BareItem, Decimal, FromPrimitive};
+    /// let bare_item = BareItem::Boolean(true.into());
+    /// assert_eq!(bare_item.as_bool().unwrap(), true);
+    /// ```
+    pub fn as_bool(&self) -> Option<bool> {
+        match self {
+            BareItem::Boolean(val) => Some(val.0),
+            _ => None,
+        }
+    }
+    /// If `BareItem` is a `Token`, returns `&str`, otherwise returns `None`.
+    /// ```
+    /// use sfv::BareItem;
+    /// # use std::convert::TryInto;
+    /// # fn main() -> Result<(), &'static str> {
+    ///
+    /// let bare_item = BareItem::Token("*bar".try_into()?);
+    /// assert_eq!(bare_item.as_token().unwrap(), "*bar");
+    /// # Ok(())
+    /// # }
+    /// ```
+    pub fn as_token(&self) -> Option<&str> {
+        match *self {
+            BareItem::Token(ref val) => Some(val),
+            _ => None,
+        }
+    }
+}
+
+impl TryFrom<i64> for BareItem {
+    type Error = &'static str;
+    /// Converts `i64` into `BareItem::Integer`.
+    /// ```
+    /// # use sfv::BareItem;
+    /// # use std::convert::TryInto;
+    /// # fn main() -> Result<(), &'static str> {
+    /// let bare_item: BareItem = 456_i64.try_into()?;
+    /// assert_eq!(bare_item.as_int().unwrap(), 456);
+    /// # Ok(())
+    /// # }
+    /// ```
+    fn try_from(item: i64) -> Result<Self, Self::Error> {
+        Ok(BareItem::Integer(item.try_into()?))
+    }
+}
+
+impl TryFrom<rust_decimal::Decimal> for BareItem {
+    type Error = &'static str;
+    /// Converts `rust_decimal::Decimal` into `BareItem::Decimal`.
+    /// ```
+    /// # use sfv::{BareItem, FromPrimitive};
+    /// # use std::convert::TryInto;
+    /// use rust_decimal::Decimal;
+    /// # fn main() -> Result<(), &'static str> {
+    /// let decimal_number = Decimal::from_f64(48.01).unwrap();
+    /// let bare_item: BareItem = decimal_number.try_into()?;
```

Comment:
I'm starting to wonder if try_into is the most ergonomic solution.
Whouldn't it be nicer to have `let decimal_number = BareItem::decimal_from_f64(48.01)?;` - and similar for the other subtypes?

Reply:
This is kinda like 3.i. from the issue (just with a better method naming than mine):

> We can have static methods on the primitives that allow early validation, like BareItem::try_token("foo") which returns a Result of either the token or the validation error. The actual token would be "unverified" in the sense that one could still create BareItem::Token("12345").

Happy to explore that!

Reply:
Also I do think both ways can be supported. `TryFrom<f64>` could simply call `BareItem::decimal_from_f64`

Reply:
@valenting what's your stance on having the external API only accepting `f64`? Do you want to keep the usage of `rust_decimal::Decimal`? The spec seems to only allow 12 digits on the integer component and 3 digits on the fractional component. f64 should cover that completely? 🤔